core/src/test/java/org/apache/iceberg/TestSnapshotManager.java

```diff
@@ -250,4 +248,305 @@ public void testCherryPickOverwrite() {
+        () -> table.manageSnapshots().createTag("tag1", snapshotId).commit());
+
+    // Trying to create another tag within the same chain
+    AssertHelpers.assertThrows("Should fail validation check if we try and create tag again",
+        IllegalArgumentException.class, "Ref tag2 already exists",
+        () -> table.manageSnapshots()
+            .createBranch("tag2", snapshotId)
+            .createBranch("tag2", snapshotId).commit());
+  }
+
+  @Test
+  public void testRemoveBranch() {
+    table.newAppend()
+        .appendFile(FILE_A)
+        .commit();
+    long snapshotId = table.currentSnapshot().snapshotId();
+    // Test a basic case of creating and then removing a branch and tag
+    table.manageSnapshots()
+        .createBranch("branch1", snapshotId)
+        .commit();
+    table.manageSnapshots()
+        .removeBranch("branch1")
+        .commit();
+
+    TableMetadata updated = table.ops().refresh();
+    SnapshotRef expectedBranch = updated.ref("branch1");
+    Assert.assertTrue(expectedBranch == null);
+
+    // Test chained creating and removal of branch and tag
+    table.manageSnapshots()
+        .createBranch("branch2", 1)
+        .removeBranch("branch2")
+        .commit();
+    updated = table.ops().refresh();
+    Assert.assertTrue(updated.ref("branch2") == null);
+
+    AssertHelpers.assertThrows("Should fail validation check if we try to remove a non-existant branch",
+        IllegalArgumentException.class, "No such branch non-existing",
+        () -> table.manageSnapshots().removeBranch("non-existing").commit());
+
+    AssertHelpers.assertThrows("Should fail validation check if we try to remove main",
+        IllegalArgumentException.class, "Cannot remove main branch",
+        () -> table.manageSnapshots().removeBranch(SnapshotRef.MAIN_BRANCH).commit());
+  }
+
+  @Test
+  public void testRemoveTag() {
+    table.newAppend()
+        .appendFile(FILE_A)
+        .commit();
+    long snapshotId = table.currentSnapshot().snapshotId();
+    // Test a basic case of creating and then removing a branch and tag
+    table.manageSnapshots()
+        .createTag("tag1", snapshotId)
+        .commit();
+    table.manageSnapshots()
+        .removeTag("tag1")
+        .commit();
+    TableMetadata updated = table.ops().refresh();
+    SnapshotRef expectedTag = updated.ref("tag1");
+    Assert.assertTrue(expectedTag == null);
+
+    // Test chained creating and removal of a tag
+    table.manageSnapshots()
+        .createTag("tag2", 1)
+        .removeTag("tag2")
+        .commit();
+    updated = table.ops().refresh();
+    Assert.assertTrue(updated.ref("tag2") == null);
+    AssertHelpers.assertThrows("Should fail validation check if we try to remove a non-existant branch",
+        IllegalArgumentException.class, "No such tag non-existing",
+        () -> table.manageSnapshots().removeTag("non-existing").commit());
+  }
+
+  @Test
+  public void testReplaceBranch() {
```

Comment:
Is there a test for fast-forward? If not I think there should be one for normal fast-forward and one to validate the ancestor check.

Reply:
Added 2 tests for fast forward for the happy path and the ancestor-check failure case